src/caffe/layers/euclidean_loss_layer.cu

```diff
@@ -6,9 +6,29 @@
 namespace caffe {
 
 template <typename Dtype>
+__global__ void EuclideanLossForwardGPU(const int n,
+          Dtype* label_data_, Dtype* bottom_data_,
+          const int ignore_label_) {
+  CUDA_KERNEL_LOOP(index, n) {
+    const int label_value = static_cast<int>(label_data_[index]);
+    if (label_value == ignore_label_) {
+      label_data_[index] = ignore_label_;
+      bottom_data_[index] = ignore_label_;
```

Comment:
Same comment: layers should not be modifying the `bottom` data in the forward pass.